array_api_strict/tests/test_elementwise_functions.py

```diff
@@ -126,33 +135,78 @@ def test_missing_functions():
     def _array_vals(dtypes):
-        for d in dtypes:
-            yield asarray(1., dtype=d)
-
-    # Use the latest version of the standard so all functions are included
-    set_array_api_strict_flags(api_version="2024.12")
-
-    for func_name, types in elementwise_function_input_types.items():
-        dtypes = _dtype_categories[types]
-        func = getattr(_elementwise_functions, func_name)
+        for dtype in dtypes:
+            yield asarray(1., dtype=dtype, device=device)
+
+    dtypes = _dtype_categories[types]
+    func = getattr(_elementwise_functions, func_name)
+
+    for x in _array_vals(dtypes):
+        if nargs(func) == 2:
+            # This way we don't have to deal with incompatible
+            # types of the two arguments.
+            r = func(x, x)
+            assert r.device == x.device
+
+        else:
+            # `atanh` needs a slightly different input value from
+            # everyone else
+            if func_name == "atanh":
+                x -= 0.1
+            r = func(x)
+            assert r.device == x.device
+
+
+@pytest.mark.parametrize("func_name", elementwise_binary_function_names)
+def test_elementwise_function_device_mismatch(func_name):
+    func = getattr(_elementwise_functions, func_name)
+    dtypes = elementwise_function_input_types[func_name]
+    if dtypes in ("floating-point", "real floating-point"):
+        dtype = float64
+    elif dtypes == "boolean":
+        dtype = xp_bool
+    else:
+        dtype = int64
+
+    a = asarray(1, dtype=dtype, device=CPU_DEVICE)
+    b = asarray(1, dtype=dtype, device=Device("device1"))
+    _ = func(a, a)
+    with pytest.raises(ValueError, match="different devices"):
+        func(a, b)
+
+
+@pytest.mark.parametrize("func_name", elementwise_function_input_types)
+def test_elementwise_function_vs_numpy_generics(func_name):
+    """
+    Test that NumPy generics are explicitly disallowed.
+
+    This must notably includes np.float64 and np.complex128, which are
+    subclasses of float and complex respectively.
+    """
+    func = getattr(_elementwise_functions, func_name)
+    dtypes = elementwise_function_input_types[func_name]
+    xp_dtypes = _dtype_categories[dtypes]
+    np_dtypes = [dtype._np_dtype for dtype in xp_dtypes]
+
+    value = 0.5 if func_name == "atanh" else 1
+    for xp_dtype in xp_dtypes:
+        for np_dtype in np_dtypes:
+            a = asarray(value, dtype=xp_dtype, device=CPU_DEVICE)
+            b = np.asarray(value, dtype=np_dtype)[()]
 
-        for x in _array_vals(dtypes):
             if nargs(func) == 2:
-                # This way we don't have to deal with incompatible
-                # types of the two arguments.
-                r = func(x, x)
-                assert r.device == x.device
-
+                _ = func(a, a)
+                with pytest.raises(TypeError, match="neither Array nor Python scalars"):
+                    func(a, b)
```

Comment:
If we want to be really thorough, can add also test `func(b, a)`: unlike https://github.com/data-apis/array-api-strict/pull/103, functions we control.